receiver/splunkhecreceiver/receiver.go

```diff
@@ -0,0 +1,319 @@
+	"go.opentelemetry.io/collector/translator/conventions"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/common/splunk"
+)
+
+const (
+	defaultServerTimeout = 20 * time.Second
+
+	responseOK                 = "OK"
+	responseInvalidMethod      = `Only "POST" method is supported`
+	responseInvalidContentType = "\"Content-Type\" must be \"application/json\""
+	responseInvalidEncoding    = "\"Content-Encoding\" must be \"gzip\" or empty"
+	responseErrGzipReader      = "Error on gzip body"
+	responseErrReadBody        = "Failed to read message body"
+	responseErrUnmarshalBody   = "Failed to unmarshal message body"
+	responseErrNextConsumer    = "Internal Server Error"
+
+	// Centralizing some HTTP and related string constants.
+	jsonContentType           = "application/json"
+	gzipEncoding              = "gzip"
+	httpContentTypeHeader     = "Content-Type"
+	httpContentEncodingHeader = "Content-Encoding"
+)
+
+var (
+	errNilNextConsumer = errors.New("nil nextConsumer")
+	errEmptyEndpoint   = errors.New("empty endpoint")
+
+	okRespBody               = initJSONResponse(responseOK)
+	invalidMethodRespBody    = initJSONResponse(responseInvalidMethod)
+	invalidContentRespBody   = initJSONResponse(responseInvalidContentType)
+	invalidEncodingRespBody  = initJSONResponse(responseInvalidEncoding)
+	errGzipReaderRespBody    = initJSONResponse(responseErrGzipReader)
+	errReadBodyRespBody      = initJSONResponse(responseErrReadBody)
+	errUnmarshalBodyRespBody = initJSONResponse(responseErrUnmarshalBody)
+	errNextConsumerRespBody  = initJSONResponse(responseErrNextConsumer)
+)
+
+// splunkReceiver implements the component.MetricsReceiver for Splunk HEC metric protocol.
+type splunkReceiver struct {
+	sync.Mutex
+	logger       *zap.Logger
+	config       *Config
+	nextConsumer consumer.MetricsConsumer
+	server       *http.Server
+
+	startOnce sync.Once
+	stopOnce  sync.Once
+}
+
+var _ component.MetricsReceiver = (*splunkReceiver)(nil)
+
+// New creates the Splunk HEC receiver with the given configuration.
+func New(
+	logger *zap.Logger,
+	config Config,
+	nextConsumer consumer.MetricsConsumer,
+) (component.MetricsReceiver, error) {
+	if nextConsumer == nil {
+		return nil, errNilNextConsumer
+	}
+
+	if config.Endpoint == "" {
+		return nil, errEmptyEndpoint
+	}
+
+	r := &splunkReceiver{
+		logger:       logger,
+		config:       &config,
+		nextConsumer: nextConsumer,
+		server: &http.Server{
+			Addr: config.Endpoint,
+			// TODO: Evaluate what properties should be configurable, for now
+			//		set some hard-coded values.
+			ReadHeaderTimeout: defaultServerTimeout,
+			WriteTimeout:      defaultServerTimeout,
+		},
+	}
+
+	return r, nil
+}
+
+// StartMetricsReception tells the receiver to start its processing.
+// By convention the consumer of the received data is set when the receiver
+// instance is created.
+func (r *splunkReceiver) Start(_ context.Context, host component.Host) error {
+	r.Lock()
```

Comment:
Is this needed due to Serve() being called in separate goroutine from the Close() call or some other reason?

Reply:
I copied this from sfxreceiver. Seems like that changed since, at the very least the mutex presents starting and stopping at the same time. Not sure it's needed now.

Reply:
It is safe to have it because I don't think there is any guarantee that the Start and Shutdown methods are always called from the same goroutine, so at the very least the mutex forces memory ordering to be right, even if they aren't called in parallel.

You don't need it for the Serve() call because goroutine startup is a guaranteed memory barrier in the Go memory model.